### ISSUE TITLE
Make the generator function example clearer

### ISSUE THREAD

**Issue by merelcht** [kedro-org/kedro#2403]
## Description
After the release announcement a user came to me and asked about the generator function feature. The user struggled understanding how to use it with a dataset other than the `MemoryDataSet` as the example shows: https://docs.kedro.org/en/stable/nodes_and_pipelines/nodes.html#how-to-use-generator-functions-in-a-node

## Context
We should improve the docs so any user can follow and implement generator functions in a node:
- the logs should show the custom dataset
- the custom dataset should be much simpler, e.g. extending from pandas.CSVDataSet and only modifying the _save method, so people actually understand what is the actual difference between the datasets

**Comment by noklam** [comment]
I am keen to take this one. I was testing this feature today. The current docs missed some catalog entries so the "yield" `_save` method wasn't tested.

I have some questions which need more discussion or go into tech design.
* How does this work with `PartitionedDataSet`? I feel the chunkwise processing has a similar pattern, should users pick one of them or they can work well together?

I hope I will gather more information while working on this.

**Comment by stichbury** [comment]
@noklam Where did this work lead you to? Is there anything new to add as a docs ticket or prioritise as updates to docs?

**Comment by stichbury** [comment]
@noklam I added you to this to ensure I don't take it out of the sprint if you want to tackle it. I don't have the bandwidth for this docs ticket so will leave you to decide.

**Comment by noklam** [comment]
https://www.linen.dev/s/kedro/t/13227438/hey-there-hope-you-are-all-doing-well-some-open-ended-questi is related too

**Comment by noklam** [comment]
I think this may related to https://github.com/kedro-org/kedro-devrel/issues/49. I think `generator` is not an easy concept and we need to explain it so that people can understand why it could be useful.

We put this in `Use generator in Node`, but it actually requires a custom Dataset + using `yield` in node, which are all advance features.

I need to go through the original PR to understand this better first.

**Comment by noklam** [comment]
I am revising the estimation of this work - it's more than I thought because of various issues. i.e. The original example does not work at all - so it's not just updating to make it clearer but a rework of the example.

**Comment by noklam** [comment]
I have discussed with @stichbury and this will be a 2-part works.
- #2858 will only be fixing the documentation
- A separate ticket to re-organise the documentations

Some discussed point @stichbury (feel free to edit my comments and make this an issue)
- The example should not live in `Node`, we discussed similar structure to [Configuration](https://docs.kedro.org/en/latest/configuration/index.html)
- Potentially a page for advance use case which are **performance related**. It shouldn't be all new contents but rather links and pointers to the corresponding features.
  - Many ideas for this [Performance]https://github.com/kedro-org/kedro-devrel/issues/49 
- Explains WHY people need generators - key points here IMO is `lazy loading and lazy saving`. Explaining how generator works maybe out of scope.

This should address questions similiar to this
![image](https://github.com/kedro-org/kedro/assets/18221871/17c87caa-ee2d-46be-82d1-00d9636620ad)
